diff -cr compface/Makefile /home/karsten/src/Projects/M/compface/Makefile
*** compface/Makefile	Tue Mar 21 19:09:40 1995
--- /home/karsten/src/Projects/M/compface/Makefile	Sat Mar  7 14:26:18 1998
***************
*** 39,45 ****
  HEADERS		= compface.h
  MAN1DIR		= $(MANDIR)/man1
  MAN3DIR		= $(MANDIR)/man3
! OBJECTS		= arith.o file.o compress.o gen.o uncompface.o
  SOURCES		= compface.c uncompface.o arith.c file.c compress.c gen.c \
  		  cmain.c uncmain.c
  HDRS		= compface.h data.h
--- 39,45 ----
  HEADERS		= compface.h
  MAN1DIR		= $(MANDIR)/man1
  MAN3DIR		= $(MANDIR)/man3
! OBJECTS		= arith.o file.o compress.o gen.o uncompface.o compface.o
  SOURCES		= compface.c uncompface.o arith.c file.c compress.c gen.c \
  		  cmain.c uncmain.c
  HDRS		= compface.h data.h
diff -cr compface/compface.c /home/karsten/src/Projects/M/compface/compface.c
*** compface/compface.c	Wed Jun 19 07:29:12 1991
--- /home/karsten/src/Projects/M/compface/compface.c	Sat Mar  7 14:29:18 1998
***************
*** 13,19 ****
   *  to me, then an attempt will be made to fix them.
   */
  
! #define MAIN
  
  #include "compface.h"
  
--- 13,19 ----
   *  to me, then an attempt will be made to fix them.
   */
  
! /*#define MAIN*/
  
  #include "compface.h"
  
