diff -ru imap-4.4.orig/src/c-client/pop3.c ../imap-4.4.orig/src/c-client/pop3.c
--- imap-4.4.orig/src/c-client/pop3.c	Fri Aug 28 02:08:39 1998
+++ ../imap-4.4.orig/src/c-client/pop3.c	Fri Jul 23 16:34:11 1999
@@ -570,7 +570,7 @@
   if ((flags & FT_UID) && !(msgno = mail_msgno (stream,msgno))) return NIL;
 				/* have header text? */
   if (!(elt = mail_elt (stream,msgno))->private.msg.header.text.data) {
-    elt->private.msg.header.text.size = pop3_cache (stream,elt);
+    elt->private.msg.header.text.size = pop3_cache_header (stream,elt);
 				/* read the header */
     fread (elt->private.msg.header.text.data = (unsigned char *)
 	   fs_get ((size_t) elt->private.msg.header.text.size + 1),
@@ -616,7 +616,7 @@
 unsigned long pop3_cache (MAILSTREAM *stream,MESSAGECACHE *elt)
 {
 				/* already cached? */
-  if (LOCAL->msgno != elt->msgno) {
+  if (LOCAL->msgno != elt->msgno || elt->rfc822_size == 1) {
 				/* no, close current file */
     if (LOCAL->txt) fclose (LOCAL->txt);
     LOCAL->txt = NIL;
@@ -626,6 +626,37 @@
 				/* load the cache */
       LOCAL->txt = netmsg_slurp (LOCAL->netstream,&elt->rfc822_size,
 				 &LOCAL->hdrsize);
+    }
+    else elt->deleted = T;
+  }
+  return LOCAL->hdrsize;
+}
+
+/* POP3 cache message header only
+ * Accepts: mail stream
+ *	    message number
+ * Returns: header size
+ */
+
+unsigned long pop3_cache_header (MAILSTREAM *stream,MESSAGECACHE *elt)
+{
+				/* already cached? */
+  if (LOCAL->msgno != elt->msgno) {
+    char tmp[MAILTMPLEN];
+				/* no, close current file */
+    if (LOCAL->txt) fclose (LOCAL->txt);
+    LOCAL->txt = NIL;
+    LOCAL->msgno = LOCAL->hdrsize = 0;
+    sprintf (tmp,"%lu 0",mail_uid (stream,elt->msgno));
+    if (pop3_send (stream,"TOP",tmp)) {
+      LOCAL->msgno = elt->msgno;/* set as current message number */
+				/* load the cache */
+      LOCAL->txt = netmsg_slurp (LOCAL->netstream,&elt->rfc822_size,
+				 &LOCAL->hdrsize);
+      elt->rfc822_size = 1;
+      /* because we don't really know, but we want mail_fetch_structure
+	 to think that it has enough info without calling dtb->text */
+      printf("rfc822_size = %d\n", elt->rfc822_size);
     }
     else elt->deleted = T;
   }
diff -ru imap-4.4.orig/src/c-client/pop3.h ../imap-4.4.orig/src/c-client/pop3.h
--- imap-4.4.orig/src/c-client/pop3.h	Wed Aug 19 19:37:37 1998
+++ ../imap-4.4.orig/src/c-client/pop3.h	Fri Jul 23 15:36:45 1999
@@ -81,6 +81,7 @@
 		   long flags);
 long pop3_text (MAILSTREAM *stream,unsigned long msgno,STRING *bs,long flags);
 unsigned long pop3_cache (MAILSTREAM *stream,MESSAGECACHE *elt);
+unsigned long pop3_cache_header (MAILSTREAM *stream,MESSAGECACHE *elt);
 long pop3_ping (MAILSTREAM *stream);
 void pop3_check (MAILSTREAM *stream);
 void pop3_expunge (MAILSTREAM *stream);

