--- imap-4.5/src/c-client/rfc822.c	Fri Sep 11 06:58:48 1998
+++ c-client/rfc822.c	Wed Feb 17 21:20:51 1999
@@ -109,6 +109,29 @@
  *	    message body
  */
 
+static char **rfc822_extraheaders_names = NULL;
+static char **rfc822_extraheaders_values = NULL;
+
+void rfc822_setextraheaders(char **names, char **values)
+{
+   rfc822_extraheaders_names = names;
+   rfc822_extraheaders_values = values;
+}
+
+void rfc822_extraheaders_lines (char **header,ENVELOPE *env)
+{
+   char **names = rfc822_extraheaders_names;
+   char **values = rfc822_extraheaders_values;
+   while(names && *names)
+   {
+      if (values) sprintf ((*header += strlen (*header)),"%s%s: %s\015\012",
+                           env->remail ? "ReSent-" : "",
+                           *(names++),
+                           *(values++)
+         );
+   }
+}
+
 void rfc822_header (char *header,ENVELOPE *env,BODY *body)
 {
   if (env->remail) {		/* if remailing */
@@ -135,6 +158,9 @@
   rfc822_header_line (&header,"Message-ID",env,env->message_id);
   rfc822_header_line (&header,"Followup-to",env,env->followup_to);
   rfc822_header_line (&header,"References",env,env->references);
+
+  rfc822_extraheaders_lines (&header,env);
+  
   if (body && !env->remail) {	/* not if remail or no body structure */
     strcat (header,"MIME-Version: 1.0\015\012");
     rfc822_write_body_header (&header,body);
